### PR TITLE
ARROW-1531: [C++] Return ToBytes by value from Decimal128

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Review by wesm — COMMENTED** [review]
I don't think this has any performance implementations, does it? The downside of this change is that you can no longer pass l-values to this function

**Comment by cpcloud** [comment]
You can still pass them: http://en.cppreference.com/w/cpp/utility/forward

> * If a call to wrapper() passes an rvalue std::string, then T is deduced to std::string (not std::string&, const std::string&, or std::string&&), and std::forward ensures that an rvalue reference is passed to foo. 
> * If a call to wrapper() passes a const lvalue std::string, then T is deduced to const std::string&, and std::forward ensures that a const lvalue reference is passed to foo.
> * If a call to wrapper() passes a non-const lvalue std::string, then T is deduced to std::string&, and std::forward ensures that a non-const lvalue reference is passed to foo.

**Comment by wesm** [comment]
I mean that this code no longer works:

```c++

FixedSizeBinaryBuilder builder(type);

std<array, SIZE> val = ...;

builder.Append(val);
```

You would have to do `builder.Append(std::move(val));`. Might be wrong, should try compiling it before I put my foot in my mouth =)

**Comment by cpcloud** [comment]
That should still work, because that's a non-const lvalue which would be deduced to

```
std::array<uint8_t, 16>&
```

**Comment by wesm** [comment]
I tried compiling:

```c++
TEST_F(TestFWBinaryArray, Builder2) {
  InitBuilder(8);

  std::array<uint8_t, 8> val;

  ASSERT_OK(builder_->Append<8>(val));
}
```

```c++
../src/arrow/array-test.cc:1207:39: error: no matching member function for call to 'Append'
  do { ::arrow::Status s = (builder_->Append<8>(val)); if (!s.ok()) { return ::testing::internal::AssertHelper(::testing::TestPartResult::kFatalFailure, "../src/arrow/array-test.cc", 1207, "Failed") = ::testing::Message() << s.ToString(); } } while (0);
                            ~~~~~~~~~~^~~~~~~~~
../src/arrow/builder.h:743:10: note: candidate function [with NBYTES = 8] not viable: no known conversion from 'std::array<uint8_t, 8>' (aka 'array<unsigned char, 8>') to 'std::array<uint8_t, 8UL> &&' (aka 'array<unsigned char, 8UL> &&') for 1st argument
  Status Append(std::array<uint8_t, NBYTES>&& value) {
```

**Comment by wesm** [comment]
Can you change the title? I'm going to cut the 0.7.0 RC after these tests pass and I can merge this

File: cpp/src/arrow/util/decimal.h [hunk 1]
```diff
@@ -104,17 +104,15 @@ class ARROW_EXPORT Decimal128 {
   /// \brief Get the low bits of the two's complement representation of the number.
   uint64_t low_bits() const { return low_bits_; }
 
-  /// \brief Put the raw bytes of the value into a pointer to uint8_t.
-  Status ToBytes(std::array<uint8_t, 16>* out) const;
+  /// \brief Return the raw bytes of the value.
+  std::array<uint8_t, 16> ToBytes() const;
```
Comment (wesm) [inline cpp/src/arrow/util/decimal.h]: It might be worth having a void ToBytes(uint8_t* out) const version, too, for writing into preallocated memory

**Comment by cpcloud** [comment]
done

**Comment by wesm** [comment]
Appveyor build in progress, will merge once that's looking good: https://ci.appveyor.com/project/cpcloud/arrow/build/1.0.286

**Comment by cpcloud** [comment]
Appveyor build is green: https://ci.appveyor.com/project/cpcloud/arrow/build/1.0.286

**Comment by cpcloud** [comment]
I rebased this PR since then, just FYI

**Pull request closed without merging**